### PR TITLE
Update format parsing to support f-strings and the str.format function

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

This PR takes over https://github.com/PyCQA/pylint/pull/6555 in hopes of getting the work merged.

Closes https://github.com/PyCQA/pylint/issues/6163
Closes https://github.com/PyCQA/pylint/issues/6085

### CODE REVIEW

**Comment by DanielNoord** [comment]
@clavedeluna Will you also work on fixing the tests? Otherwise it doens't make a lot of sense to keep this PR open.

**Comment by clavedeluna** [comment]
> @clavedeluna Will you also work on fixing the tests? Otherwise it doens't make a lot of sense to keep this PR open.

Of course. What would be the point of this PR otherwise? The following work needs to happen:

- [ ] add typing
- [ ] fix tests

Typing is the first thing I think but I haven't had the energy to work on it much yet. If anyone is interested in contributing, please feel free to push commits.

**Comment by DanielNoord** [comment]
I'd focus on the tests first. The original PR never got the test suite to pass so there might be crucial incompatible changes in here. Typing of those change might become redundant after the incompatibilities have been fixed.

**Comment by clavedeluna** [comment]
I can now see why this was so difficult to get right in the first place. I did quite a bit but there are two tests that fail, and in trying to fix them so many other things break.

@DanielNoord maybe I won't be able to complete this, but I did find a couple functional tests that I'm not sure are right. In string_formatting.py:

```
"{0.missing.length}".format(ReturnYes())
 "{1.missing.length}".format(ReturnYes())
```

shouldn't those two lines emit some message, given that `missing` is not a real module?

**Comment by DanielNoord** [comment]
@clavedeluna Which messages were you thinking of?

**Comment by clavedeluna** [comment]
> @clavedeluna Which messages were you thinking of?

Given this
```
"{0.missing}".format(2) # [missing-format-attribute]
```

I would expect

```
from missing import Missing
class ReturnYes:
    """ can't be properly inferred """
    missing = Missing()


"{0.missing.length}".format(ReturnYes())
"{1.missing.length}".format(ReturnYes())
```

to also raise `missing-format-attribute`. In the first example, `2` doesn't have a `.missing` attr. In the second group of cases, `ReturnYes().missing.length` also doesn't exit.

**Comment by DanielNoord** [comment]
Sorry for the late response, but I agree. That seems to make sense to me.

File: pylint/checkers/strings.py [hunk 1]
```diff
@@ -404,10 +424,75 @@ def visit_binop(self, node: nodes.BinOp) -> None:
...
+                    continue
+                try:
+                    arg_type = utils.safe_infer(field)
+                except astroid.InferenceError:
+                    continue
+
+                if len(value.format_spec.values) == 0:
+                    continue
+
+                def convert_node_to_string(x):
```
Comment (clavedeluna) [inline pylint/checkers/strings.py:469]: move this func out for performance

File: pylint/checkers/strings.py [hunk 2]
```diff
@@ -449,6 +534,106 @@ def _detect_vacuous_formatting(
...
+                format_type = conversion
+
+            if (
+                format_type is not None
+                and arg_type
+                and arg_type != astroid.Uninferable
+                and not arg_matches_format_type(arg_type, format_type)
+            ):
+                self.add_message(
+                    "bad-string-format-type",
```
Comment (clavedeluna) [inline pylint/checkers/strings.py:632]: try to dedupe with above code

File: pylint/checkers/strings.py [hunk 3]
```diff
@@ -634,6 +846,51 @@ def _check_new_format_specifiers(
                     # can't check further if we can't infer it
                     break
 
+        for position, specifiers in pos_args.items():
```
Comment (clavedeluna) [inline pylint/checkers/strings.py:849]: try to dedupe with above code

**Comment by clavedeluna** [comment]
I'm thinking that the two primer reports are false positives but the one in music21 seems that it may be really hard to fix because `:d` is referring to the return from a function

**Comment by Pierre-Sassoulas** [comment]
I think for music21 bools can be considered as ``int`` too so it should not raise even without considering the function return, right ?

**Comment by clavedeluna** [comment]
I'm going to open this PR up for taking over since I just can't find the time to come back to it. Here is what I believe remains to be done:
- fix primer results as per Pierre's comment above
- fix pre-commit. this will require adding typing

**Comment by DanielNoord** [comment]
Closing due to inactivity. Sorry this seems to have died due to me not responding actively enough. Life caught up to me.

**Pull request closed without merging**